modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryManager.java

```diff
@@ -102,29 +110,43 @@ public QueryMemoryManager(GridKernalContext ctx, long globalQuota) {
 
         this.blockSize = Long.getLong(IgniteSystemProperties.IGNITE_SQL_MEMORY_RESERVATION_BLOCK_SIZE, DFLT_MEMORY_RESERVATION_BLOCK_SIZE);
         this.globalQuota = globalQuota;
-        this.dfltSqlQryMemoryLimit = dfltMemLimit;
+        // TODO GG-18629 - get from configuration.
+        this.failOnMemLimitExceed = !Boolean.getBoolean(IgniteSystemProperties.IGNITE_SQL_USE_DISK_OFFLOAD);
 
-        this.reserveOp = new ReservationOp(globalQuota);
+        this.dfltSqlQryMemoryLimit = dfltMemLimit;
 
         this.log = ctx.log(QueryMemoryManager.class);
 
         metrics = new SqlStatisticsHolderMemoryQuotas(this, ctx.metric());
     }
 
     /** {@inheritDoc} */
-    @Override public void reserve(long size) {
+    @Override public boolean reserved(long size) {
         if (size == 0)
-            return; // Nothing to do.
+            return true; // Nothing to do.
 
-        assert size > 0;
+        long reserved0 = reserved.addAndGet(size);
```

Comment:
Can we use GridAtomicLong.checkAndSet() here.
A predicate can fails with exception if offload is disabled (failOnMemLimitExceed==false) as it works before,
and return true\false if offload is enabled